### PR TITLE
Add secret key Stream

### LINKED ISSUE #141: Add SecretStream API to `sel`
Let's add a high-level interface to https://hackage.haskell.org/package/libsodium-bindings/docs/LibSodium-Bindings-SecretStream.html in `sel`.

@domenkozar I would love to hear about your use-case, this will help us define the initial scope and write tests. :)

### PR DESCRIPTION
**Pull request by Kleidukos**
closes #141

### CODE REVIEW

**Comment by domenkozar** [comment]
Nonce is managed by the lower level?

**Comment by Kleidukos** [comment]
@domenkozar Absolutely, as said here: https://doc.libsodium.org/secret-key_cryptography/secretstream#algorithm

> Initialization (secretstream_init): a subkey k and a 64-bit nonce n are derived from a key K and a 192-bit random nonce N, using the same algorithm as XChaCha20

libsodium handles this for us. :)

**Comment by Kleidukos** [comment]
Okay, here is the provisional API:

Screenshot 2024-02-16 at 11-57-55 Sel SecretKey Stream

**Review by kozross — APPROVED** [review]

File: sel/src/Sel/SecretKey/Stream.hs [hunk 1]
```diff
@@ -0,0 +1,571 @@
...
+import Control.Exception (Exception, throw)
+import Control.Monad (forM, when)
+import Control.Monad.IO.Class (MonadIO, liftIO)
+import Data.Base16.Types (Base16)
+import qualified Data.Base16.Types as Base16
+import Data.ByteString (StrictByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Base16 as Base16
+import qualified Data.ByteString.Internal as BS
+import qualified Data.ByteString.Unsafe as BS
```
[the remark this answers is not in the document]
Comment (Kleidukos) [reply sel/src/Sel/SecretKey/Stream.hs:76]: In this case the packages `bytestring` is imported under BS and the package `base16` is imported under `Base16`. but perhaps you meant "module namespace"?

[the remark this answers is not in the document]
Comment (kozross) [reply sel/src/Sel/SecretKey/Stream.hs:76]: `Data.Bytestring`, `Data.ByteString.Internal`, and `Data.ByteString.Unsafe` are all imported as `BS`. That's what I'm unhappy about.

File: sel/src/Sel/SecretKey/Stream.hs [hunk 2]
```diff
@@ -0,0 +1,571 @@
...
+        (messageTagToConstant messageTag)
+    when (result /= 0) $ throw StreamEncryptionException
+  pure $ CipherText (fromIntegral cStringLen) cipherTextFPtr
+
+-- | Perform streaming encryption of a finite Linked List.
+--
+-- This function can throw 'StreamEncryptionException' upon an error in the underlying implementation.
+--
+-- @since 0.0.1.0
+encryptList :: forall m. MonadIO m => SecretKey -> [StrictByteString] -> m (Header, [CipherText])
```
[the remark this answers is not in the document]
Comment (kozross) [reply sel/src/Sel/SecretKey/Stream.hs:211]: Ah yeah, I see that now. You could possibly rewrite this as a `foldM` combined with a list pattern match, but it's probably not clearer that way.

File: sel/src/Sel/SecretKey/Stream.hs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kleidukos) [reply sel/src/Sel/SecretKey/Stream.hs:76]: ah so not packages but namespaces. Got it! :+1:

**Review by kozross — APPROVED** [review]

File: sel/src/Sel/SecretKey/Stream.hs [hunk 3]
```diff
@@ -0,0 +1,571 @@
...
+-- | Create a 'SecretKey' from a binary 'StrictByteString' that you have obtained on your own,
+-- usually from the network or disk.
+--
+-- The input secret key, once decoded from base16, must be of length
+-- 'cryptoSecretStreamXChaCha20Poly1305KeyBytes'.
+--
+-- @since 0.0.1.0
+secretKeyFromHexByteString :: Base16 StrictByteString -> Either Text SecretKey
+secretKeyFromHexByteString hexSecretKey = unsafeDupablePerformIO $
+  case Base16.decodeBase16Untyped (Base16.extractBase16 hexSecretKey) of
```
Comment (intricate) [inline sel/src/Sel/SecretKey/Stream.hs:345]: Sorry for digging this up from the past, but I'm a bit confused by this and just wanted to see if there's some context I'm missing.

Why do we accept a `Base16 StrictByteString` just to call `extractBase16` and try to decode it again?

If the value is a `Base16 a`, shouldn't we assume that it's already Base16 encoded? Either that, or the user knows what they're doing and constructed the value with `assertBase16`.

Comment (intricate) [inline sel/src/Sel/SecretKey/Stream.hs:345]: Sorry for digging this up from the past, but I'm a bit confused by this and just wanted to see if there's some context I'm missing.

Why do we call `decodeBase16Untyped . extractBase16` on a `Base16 StrictByteString` when we could use `decodeBase16`?

If the value is a `Base16 a`, shouldn't we assume that it's already Base16-encoded and we can use `decodeBase16`?

Comment (Kleidukos) [reply sel/src/Sel/SecretKey/Stream.hs:345]: You're right, this is weird. I'll go over it later. Thanks!

**Pull request merged**